### PR TITLE
Added 4 modules which deal with Data independent of the ledger 

### PR DESCRIPTION
**Pull request by TimSheard**
No functions were changed in this PR, only moved to new places.
Four new modules were created in semantics/executable-spec/src/Data/  
These can be imported using the module names
Data.CannonicalMaps.hs
Data.Coders
Data.MemoBytes
Control.SetAlgebra

All changes fall into the following categories.

The file shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/MemoBytes.hs is split into two files.
     semantics/executable-spec/src/Data/Coders
     semantics/executable-spec/src/Data/MemoBytes

The file shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/ValueInternal.hs  is split into two files
     shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/ValueInternal.hs
     semantics/executable-spec/src/Data/CannonicalMaps

The file semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs  is moved to
     semantics/executable-spec/src/Control/SetAlgebra.hs

The file semantics/executable-spec/src/Control/Iterate/SetAlgebraInternal.hs  is moved to
     semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs

Several functions from shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs were moved to
     semantics/executable-spec/src/Data/Coders  for dependency reasons

A few files had to have some of their imports changed.

The main interface to the SetAlgebra library is now at Control.SetAlgebra

### CODE REVIEW

File: semantics/executable-spec/small-steps.cabal [hunk 1]
```diff
@@ -32,19 +32,24 @@ library
                      , Control.State.Transition.Simple
                      , Data.AbstractSize
                      , Data.Relation
+                     , Data.CannonicalMaps
+                     , Data.MemoBytes
+                     , Data.Coders
+                     , Data.SetAlgebraInternal
```
Comment (nc6) [inline semantics/executable-spec/small-steps.cabal]: Why `Data.SetAlgebraInternal` but `Control.SetAlgebra`? We should keep the same prefix...

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/ValueInternal.hs [hunk 2]
```diff
@@ -121,106 +125,7 @@ instance Era era => Val (Value era) where
       addrHashLen :: Integer
       addrHashLen = 28
 
--- =====================================================================================
```
Comment (nc6) [inline shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/ValueInternal.hs:124]: Given the `CanonicalMap` stuff has moved, can we now just move the core value definition into `Value` and drop the internal module?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs [hunk 3]
```diff
@@ -56,31 +56,40 @@ import Cardano.Binary
...
+    decodeList,
+    decodeRecordNamed,
+    decodeRecordSum,
+    decodeSeq,
+    decodeSet,
+    decodeStrictSeq,
+    encodeFoldable,
+    encodeFoldableEncoder,
+    wrapCBORArray,
+  )
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs:92]: If we're extracting these, maybe better to add them to cardano-binary (https://github.com/input-output-hk/cardano-base/tree/master/binary/src/Cardano/Binary). Check with @redxaxder

File: semantics/executable-spec/src/Data/MemoBytes.hs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- | MemoBytes is an abstration for a datetype that encodes its own seriialization.
```
Comment (nc6) [inline semantics/executable-spec/src/Data/MemoBytes.hs]:
```suggestion
-- | MemoBytes is an abstration for a datetype that encodes its own serialization.
```

File: semantics/executable-spec/src/Data/Coders.hs [hunk 5]
```diff
@@ -40,77 +35,126 @@ module Shelley.Spec.Ledger.MemoBytes
     encodeSeq,
     encodeStrictSeq,
     encodeSet,
-    showMemo,
-    printMemo,
+     runE,              -- for testing
```
Comment (nc6) [inline semantics/executable-spec/src/Data/Coders.hs]: [minor] Weird formatting...

File: semantics/executable-spec/small-steps.cabal (hunk 1 above)
Comment (TimSheard) [reply semantics/executable-spec/small-steps.cabal]: So now the actual code is in
Control/Iterate/SetAlgebra,hs
And the interface is in
Control/SetAlgebra.hs

File: shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/ValueInternal.hs (hunk 2 above)
Comment (TimSheard) [reply shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/ValueInternal.hs:124]: Good idea. It is done.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs (hunk 3 above)
Comment (TimSheard) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs:92]: I'll check. Perhaps we can make this a separate PR.

File: semantics/executable-spec/src/Data/MemoBytes.hs (hunk 4 above)
Comment (TimSheard) [reply semantics/executable-spec/src/Data/MemoBytes.hs]: done.

File: semantics/executable-spec/src/Data/Coders.hs (hunk 5 above)
Comment (TimSheard) [reply semantics/executable-spec/src/Data/Coders.hs]: easy to fix.

**Review by JaredCorduan — APPROVED** [review]
thanks Tim, looks good to me. I'll approve now so that you aren't block, but I guess check about those exports in Shelley.Spec.Ledger.Serialization or handle it in another PR.

**Comment by JaredCorduan** [comment]
Merging now since Tim is blocked on this, but he promises to move stuff to Base soon.

**Pull request merged**